doc/tutorials/gapi/face_beautification/face_beautification.markdown

```diff
@@ -0,0 +1,500 @@
+kernels implementing) are not going to be described in this text; there will be
+only comments about debatable features implementation.
+
+Let's start at the begining of a pipline graph definition.
+
+## Face detector inference and post-processing {#gapi_fb_face_detect}
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp gar_ROI
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp fd_inf
+
+The `config::smth` variables are constants defined in the `config` namespace:
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp config
+
+After the inference, output `cv::Mat` post-processing is required: the
+face detector returns a blob with the shape [1, 1, N, 7], where N is
+the number of bounding boxes containing detected faces. Structure of an output
+for every face is the following:
+`[image_id, label, conf, x_min, y_min, x_max, y_max]`;
+all the seven elements are floating point. For our further goals we need to take
+only results with `image_id > 0` because a negative value in this field
+indicates the end of detections; also we can cut detections by the `conf` field
+to avoid detector's mistakes; so the Mat parsing looks like this code:
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp vec_ROI
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp fd_pp
+
+Some points to be mentioned about this kernel:
+
+- It takes in a `cv::Mat` of detections and a `cv::Mat` of the input frame; it
+gives out an array of ROI's where faces have been detected.
+
+- There is `cv::Mat` data parsing by the pointer on a float used.
+
+- Float numbers we've got from the Mat are between 0 and 1 and denote normalized
+coordinates; to get the real pixel coordinates we should multiply it by the
+image sizes respectively to the directions. To be fully accurate, the received
+numbers should be then rounded the right way and casted to `int` to construct
+integer Points; these operations have been wrapped in `toIntRounded()`:
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp toInt
+
+- Rectangles are constructed by two Points --- the top-left and the bottom-right
+corners.
+
+- By far the most important thing here is solving an issue that sometimes
+detector spits out coordinates located outside of the image; if we pass such an
+ROI to be processed, errors of the landmarks detector will occure. The frame box
+`borders` is created and then intersected with the face rect (by `operator&()`)
+to handle these cases and save the ROI in `outFaces` which is for sure
+inside the frame.
+
+## Facial landmarks detector inference and post-processing {#gapi_fb_landm_detect}
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp ld_inf
+
+First of all, a few words about custom types: the `Contour` type is used for
```

Comment:
You've already used `first of all` many times. Please, change it to smth else

Reply:
OK